p11-kit/list-mechanisms.c

```diff
@@ -0,0 +1,303 @@
+	if (info.flags & CKF_GENERATE)
+		printf (" generate");
+	if (info.flags & CKF_GENERATE_KEY_PAIR)
+		printf (" generate-key-pair");
+	if (info.flags & CKF_WRAP)
+		printf (" wrap");
+	if (info.flags & CKF_UNWRAP)
+		printf (" unwrap");
+	if (info.flags & CKF_DERIVE)
+		printf (" derive");
+	if (info.flags & CKF_EXTENSION)
+		printf (" extension");
+
+	if (info.ulMaxKeySize)
+		printf (" key-size=%ld-%ld", info.ulMinKeySize, info.ulMaxKeySize);
+	printf ("\n");
+}
+
+static int
+list_mechanisms (const char *token_str)
+{
+	int ret = 1;
+	CK_FUNCTION_LIST **modules = NULL;
+	CK_FUNCTION_LIST *module = NULL;
+	P11KitUri *uri = NULL;
+	P11KitIter *iter = NULL;
+	CK_SESSION_HANDLE session = 0;
+	CK_SLOT_ID slot = 0;
+	CK_MECHANISM_INFO mechanism_info;
+	CK_MECHANISM_TYPE_PTR mechanisms = NULL;
+	CK_MECHANISM_TYPE mechanism = CKP_INVALID_ID;
+	unsigned long mechanisms_count = 0;
+	unsigned long i;
+	p11_list_printer printer;
+	CK_RV rv;
+
+	uri = p11_kit_uri_new ();
+	if (uri == NULL) {
+		p11_message (_("failed to allocate memory"));
+		goto cleanup;
+	}
+
+	if (p11_kit_uri_parse (token_str, P11_KIT_URI_FOR_OBJECT_ON_TOKEN, uri) != P11_KIT_URI_OK) {
+		p11_message (_("failed to parse URI"));
+		goto cleanup;
+	}
+
+	modules = p11_kit_modules_load_and_initialize (0);
+	if (modules == NULL) {
+		p11_message (_("failed to load and initialize modules"));
+		goto cleanup;
+	}
+
+	iter = p11_kit_iter_new (uri, P11_KIT_ITER_WITH_LOGIN);
+	if (iter == NULL) {
+		p11_message (_("failed to initialize iterator"));
+		goto cleanup;
+	}
+
+	p11_list_printer_init (&printer, stdout, 0);
+
+	uri = p11_kit_uri_new ();
+	if (uri == NULL) {
+		p11_message (_("failed to allocate memory"));
+		goto cleanup;
+	}
+
+	if (p11_kit_uri_parse (token_str, P11_KIT_URI_FOR_TOKEN, uri) != P11_KIT_URI_OK) {
+		p11_message (_("failed to parse URI"));
+		goto cleanup;
+	}
+
+	modules = p11_kit_modules_load_and_initialize (0);
+	if (modules == NULL) {
+		p11_message (_("failed to load and initialize modules"));
+		goto cleanup;
+	}
+
+	iter = p11_kit_iter_new (uri, P11_KIT_ITER_WITH_LOGIN | P11_KIT_ITER_WITH_TOKENS | P11_KIT_ITER_WITHOUT_OBJECTS);
+	if (iter == NULL) {
+		p11_message (_("failed to initialize iterator"));
+		goto cleanup;
+	}
+
+	p11_list_printer_init (&printer, stdout, 0);
+	p11_kit_iter_begin (iter, modules);
+	while ((rv = p11_kit_iter_next (iter)) == CKR_OK) {
+		module = p11_kit_iter_get_module (iter);
+		if (module == NULL) {
+			p11_message (_("failed to obtain module"));
+			goto cleanup;
+		}
+
+		slot = p11_kit_iter_get_slot (iter);
+		if (slot == 0) {
+			p11_message (_("failed to obtain slot"));
+			goto cleanup;
+		}
+		p11_list_printer_start_section (&printer, "Token", "#%lu", slot);
+
+		rv = module->C_GetMechanismList (slot, NULL, &mechanisms_count);
+		if (rv != CKR_OK) {
+			p11_message (_("querying amount of mechanisms failed: %s"), p11_kit_strerror (rv));
+			goto cleanup;
+		}
+
+		mechanisms = malloc(sizeof(CK_MECHANISM_TYPE) * mechanisms_count);
```

Comment:
```suggestion
		mechanisms = reallocarray (NULL, mechanisms_count, sizeof (CK_MECHANISM_TYPE));
```
Maybe it would also be possible to reuse the previously allocated `mechanisms` by resizing.

Reply:
right, and it was also leaking. used reallocarray